### PR TITLE
[SPARK-33420][SQL] Fix BroadCastJoin failure when keys on join side has cast from DateTyte to String

### PR DESCRIPTION
**Pull request by hzyangkai**
### What changes were proposed in this pull request?
Fix BroadCastJoin failure when keys on join side has cast from DateTyte to String

### Why are the changes needed?

1.In Expression#canonicalized, Canonicalize.execute method will be called. In it the ignoreTimeZone method will judge CastBase. If it does not need timeZone, timeZone will be forced to be removed。
```
object Canonicalize extends Logging{
    def execute(e: Expression): Expression = {
      expressionReorder(ignoreTimeZone(ignoreNamesTypes(e)))
    }
    private[expressions] def ignoreTimeZone(e: Expression): Expression = e match {
    case c: CastBase if c.timeZoneId.nonEmpty && !c.needsTimeZone =>
      logDebug(s"e:${e} ignoreTimeZone")
      c.withTimeZone(null)
    case _ => e
  }
  }
```
2.The CastBase#needsTimeZone method will call the Cast#needsTimeZone method to judge,
```
  def needsTimeZone(from: DataType, to: DataType): Boolean = (from, to) match {
            case (StringType, TimestampType | DateType) => true
            case (DateType, TimestampType) => true
            case (TimestampType, StringType) => true
            case (TimestampType, DateType) => true
            case (ArrayType(fromType, _), ArrayType(toType, _)) => needsTimeZone(fromType, toType)
            case (MapType(fromKey, fromValue, _), MapType(toKey, toValue, _)) =>
              needsTimeZone(fromKey, toKey) || needsTimeZone(fromValue, toValue)
            case (StructType(fromFields), StructType(toFields)) =>
              fromFields.length == toFields.length &&
                fromFields.zip(toFields).exists {
                  case (fromField, toField) =>
                    needsTimeZone(fromField.dataType, toField.dataType)
                }
            case _ => false
          }
```

It can be seen here that if from DateType to StringType, the judgment result is false, indicating that cast(DateTyte as String) does not require Timezone

3. But the conversion from Datetype to String in Cast requires Timezone, as following:

```
abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression with NullIntolerant {
        private lazy val dateFormatter = DateFormatter(zoneId)

        private[this] def cast(from: DataType, to: DataType): Any => Any = {
            if (DataType.equalsStructurally(from, to)) {
              identity
            } else if (from == NullType) {
              _ => throw new SparkException(s"should not directly cast from NullType to $to.")
            } else {
              to match {
                case dt if dt == from => identity[Any]
                case StringType => castToString(from)
                case BinaryType => castToBinary(from)
                case DateType => castToDate(from)
                case decimal: DecimalType => castToDecimal(from, decimal)
                case TimestampType => castToTimestamp(from)
                case CalendarIntervalType => castToInterval(from)
                case BooleanType => castToBoolean(from)
                case ByteType => castToByte(from)
                case ShortType => castToShort(from)
                case IntegerType => castToInt(from)
                case FloatType => castToFloat(from)
                case LongType => castToLong(from)
                case DoubleType => castToDouble(from)
                case array: ArrayType =>
                  castArray(from.asInstanceOf[ArrayType].elementType, array.elementType)
                case map: MapType => castMap(from.asInstanceOf[MapType], map)
                case struct: StructType => castStruct(from.asInstanceOf[StructType], struct)
                case udt: UserDefinedType[_]
                  if udt.userClass == from.asInstanceOf[UserDefinedType[_]].userClass =>
                  identity[Any]
                case _: UserDefinedType[_] =>
                  throw new SparkException(s"Cannot cast $from to $to.")
              }
            }

            private[this] def castToString(from: DataType): Any => Any = from match {
              case CalendarIntervalType =>
                buildCast[CalendarInterval](_, i => UTF8String.fromString(i.toString))
              case BinaryType => buildCast[Array[Byte]](_, UTF8String.fromBytes)
              case DateType => buildCast[Int](_, d => UTF8String.fromString(dateFormatter.format(d)))
              case TimestampType => buildCast[Long](_,
                t => UTF8String.fromString(DateTimeUtils.timestampToString(timestampFormatter, t)))
            }
        }
 trait TimeZoneAwareExpression extends Expression {
          def timeZoneId: Option[String]
          def withTimeZone(timeZoneId: String): TimeZoneAwareExpression

          @transient lazy val zoneId: ZoneId = DateTimeUtils.getZoneId(timeZoneId.get)
        }
```
4. So we should add judge in Cast#needsTimeZone as following:
```
def needsTimeZone(from: DataType, to: DataType): Boolean = (from, to) match {
            case (StringType, TimestampType | DateType) => true
            case (DateType, TimestampType) => true
            case (DateType, StringType) => true
            case (TimestampType, StringType) => true
            case (TimestampType, DateType) => true
            case (ArrayType(fromType, _), ArrayType(toType, _)) => needsTimeZone(fromType, toType)
            case (MapType(fromKey, fromValue, _), MapType(toKey, toValue, _)) =>
              needsTimeZone(fromKey, toKey) || needsTimeZone(fromValue, toValue)
            case (StructType(fromFields), StructType(toFields)) =>
              fromFields.length == toFields.length &&
                fromFields.zip(toFields).exists {
                  case (fromField, toField) =>
                    needsTimeZone(fromField.dataType, toField.dataType)
                }
            case _ => false
          }
```

### CODE REVIEW

**Comment by wangyum** [comment]
It seems https://issues.apache.org/jira/browse/SPARK-33306 has fixed this issue?

**Comment by hzyangkai** [comment]
hi, @tgravescs , could you help me review this pr please? thanks very much.

**Comment by hzyangkai** [comment]
> It seems https://issues.apache.org/jira/browse/SPARK-33306 has fixed this issue?

yes ,it is .the situation is not same,but this solvation is the same,thank you.
could you help me review SPARK-33192  pr please ?

**Comment by tgravescs** [comment]
@cloud-fan @MaxGekk who fixed the first issue

@hzyangkai  what does https://issues.apache.org/jira/browse/SPARK-33192 have to do with this issue?  Perhaps just wrong link?

**Comment by tgravescs** [comment]
can you please add a test for this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 1]
```diff
@@ -110,6 +110,7 @@ object Cast {
     case (StringType, TimestampType | DateType) => true
     case (TimestampType | DateType, StringType) => true
     case (DateType, TimestampType) => true
+    case (DateType, StringType) => true
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:113]: Why the case is not handled by the case above?
```scala
case (TimestampType | DateType, StringType) => true
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:113]: Yea it's already covered by L111. We should close this PR.

**Pull request closed without merging**